api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java

```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
+	        throws DAOException {
+		List<Patient> patients = findPatients(query, includeVoided, start, length);
+		List<Patient> result=new LinkedList<>();
+		for (Patient p : patients) {
+			if (p.getGender().equals(gender)) {
+				result.add(p);
+			}
+		}
+		return result;
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query, Date from, Date to, Integer start, Integer length,
+	        Boolean includeVoided) throws DAOException {
+		Integer tmpStart = start;
+		if (tmpStart == null) {
+			tmpStart = 0;
+		}
+		Integer maxLength = HibernatePersonDAO.getMaximumSearchResults();
+		Integer tmpLength = length;
+		if (tmpLength == null || tmpLength > maxLength) {
+			tmpLength = maxLength;
+		}
+		query = LuceneQuery.escapeQuery(query);
+
+		List<Patient> patients = new LinkedList<>();
+
+		String minChars = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_MIN_SEARCH_CHARACTERS);
+
+		if (minChars == null || !StringUtils.isNumeric(minChars)) {
+			minChars = "" + OpenmrsConstants.GLOBAL_PROPERTY_DEFAULT_MIN_SEARCH_CHARACTERS;
+		}
+		if (query.length() < Integer.valueOf(minChars)) {
+			return patients;
+		}
+		
+		PatientLuceneQuery patientLuceneQuery=new PatientLuceneQuery(sessionFactory);
+		LuceneQuery<Person> ageQuery=patientLuceneQuery.getPatinetWithAgeRange(from, to, includeVoided);
+		List<Integer> tmp=new ArrayList<>();
+		List<Object[]> tmpPatientId=ageQuery.listProjection("personId");
+		tmpPatientId.forEach(id->tmp.add((Integer)id[0]));
+
+		LuceneQuery<PatientIdentifier> identifierQuery = getPatientIdentifierLuceneQuery(query, includeVoided, false);
+
+		long identifiersSize = identifierQuery.resultSize();
+		if (identifiersSize > tmpStart) {
+			ListPart<Object[]> patientIdentifiers = identifierQuery.listPartProjection(tmpStart, tmpLength, "patient.personId");
+			patientIdentifiers.getList().forEach(patientIdentifier -> {if(tmp.contains((Integer)patientIdentifier[0])) patients.add(getPatient((Integer)patientIdentifier[0]));});
+
+			tmpLength -= patientIdentifiers.getList().size();
+			tmpStart = 0;
+		} else {
+			tmpStart -= (int) identifiersSize;
+		}
+
+		if (tmpLength == 0) {
+			return patients;
+		}
+
+		PersonLuceneQuery personLuceneQuery = new PersonLuceneQuery(sessionFactory);
+
+		LuceneQuery<PersonName> nameQuery = personLuceneQuery.getPatientNameQuery(query, includeVoided, identifierQuery);
+		
+		long namesSize = nameQuery.resultSize();
+		if (namesSize > tmpStart) {
+			ListPart<Object[]> personNames = nameQuery.listPartProjection(tmpStart, tmpLength, "person.personId");
+			personNames.getList().forEach(personName ->{if(tmp.contains((Integer)personName[0])) patients.add(getPatient((Integer) personName[0]));});
+			tmpLength -= personNames.getList().size();
+			tmpStart = 0;
+		} 
+		
+		return patients;
+
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query,Date birthdate, Integer start, Integer length,
```

Comment:
this methods body is quite similar to the above getPatients method body and have lots of duplicate code.